zone/raids.cpp

```diff
@@ -2203,3 +2253,759 @@ void Raid::SetNewRaidLeader(uint32 i)
+				slot + 1,
+				this->GetID(),
+				delegatee);
+			auto results = database.QueryDatabase(query);
+			if (!results.Success())
+				LogError("Unable to set raid main marker for player: [{}] slot: [{}] with error [{}]\n", 
+					delegatee, 
+					slot + 1, 
+					results.ErrorMessage().c_str());
+		}
+	}
+	das->DelegateAbility = RaidDelegateMainMarker;
+	das->MemberNumber = 0;
+	das->EntityID = c->GetID();
+	strcpy(das->Name, delegatee);
+	QueuePacket(outapp);
+	safe_delete(outapp);
+}
+
+int Raid::FindNextRaidDelegateSlot(int option) 
+{
+	//option 1 FindNextRaidMainMarkerSlot
+	//option 2 FindNextRaidMainAssisterSlot
+	if (option == 1) {
+		for (int i = 0; i < MAX_RAID_MAIN_MARKERS; i++) {
+			if (strlen(MainMarkerPCs[i]) == 0) {
+					return i;
+			}
+		}
+		return -1;		
+	}
+	else if (option == 2) {
+		for (int i = 0; i < MAX_RAID_MAIN_ASSISTERS; i++) {
+			if (strlen(MainAssisterPCs[i]) == 0) {
+				return i;
+			}
+		}
+		return -1;		
+	}
+}
+
+void Raid::UpdateXTargetType(XTargetType Type, Mob* m, const char* Name)
+{
+	for (const auto& rm : members) {
+		if (!rm.member || rm.is_bot || !rm.member->XTargettingAvailable())
+			continue;
+
+		for (int i = 0; i < rm.member->GetMaxXTargets(); ++i)
+		{
+			if (rm.member->XTargets[i].Type == Type)
+			{
+				if (m)
+					rm.member->XTargets[i].ID = m->GetID();
+				else
+					rm.member->XTargets[i].ID = 0;
+
+				if (Name)
+					strncpy(rm.member->XTargets[i].Name, Name, 64);
+
+				rm.member->SendXTargetPacket(i, m);
+			}
+		}
+	}
+}
+
+void Raid::RaidMarkNPC(Mob* mob, uint32 parameter)
+{
+	Client* c = mob->CastToClient();
+	if (!c || !c->GetTarget() || parameter < 1 || parameter > 3) {
+		LogDebug("RaidMarkNPC Failed sanity checks.");
+		return;
+	}
+
+	for (int i = 0; i < MAX_RAID_MAIN_MARKERS; i++) {
+		auto cname = c->GetCleanName();
+		if (strcasecmp(MainMarkerPCs[i], cname) == 0 || strcasecmp(leadername, cname) == 0) {
+			MarkedNPCs[parameter - 1] = c->GetTarget()->GetID();
+
+			std::string query = StringFormat("UPDATE raid_details SET MarkedNPC%i = %i WHERE raidid = %i;",
+				parameter,
+				MarkedNPCs[parameter - 1],
+				this->GetID());
+			auto results = database.QueryDatabase(query);
+			if (!results.Success())
+				LogError("Unable to set MarkedNPC{} from slot: [{}] with error [{}]\n",
+					parameter,
+					parameter - 1,
+					results.ErrorMessage().c_str());
+
+			auto outapp = new EQApplicationPacket(OP_MarkRaidNPC, sizeof(MarkNPC_Struct));
+			MarkNPC_Struct* mnpcs = (MarkNPC_Struct*)outapp->pBuffer;
+			mnpcs->TargetID = MarkedNPCs[parameter - 1];
+			mnpcs->Number = parameter;
+			strcpy(mnpcs->Name, c->GetTarget()->GetCleanName());
+			QueuePacket(outapp);
+			safe_delete(outapp);
+			UpdateXtargetMarkedNPC();
+			return;
+		}
+	}
+		//client is not delegated the mark ability
+		c->MessageString(Chat::Cyan, NOT_DELEGATED_MARKER);
+		return;
+}
+
+void Raid::UpdateXtargetMarkedNPC() 
+{
+	auto mm1 = entity_list.GetNPCByID(MarkedNPCs[0]);
```

Comment:
The following code can be simplified this way

```cpp
for (int i = 0; i < 3; i++) {
	auto mm = entity_list.GetNPCByID(MarkedNPCs[i]);
	if (mm) {
		UpdateXTargetType(static_cast<XTargetType>(RaidMarkTarget1 + i), mm->CastToMob(), mm->CastToMob()->GetName());
	}
	else {
		UpdateXTargetType(static_cast<XTargetType>(RaidMarkTarget1 + i), nullptr);
	}
}
```

Reply:
Updated this and a few other areas with the same pattern